### PR TITLE
Add option to write test cases to JSON file

### LINKED ISSUE #365: Option to write out all individual test case results to json
I are quite interested in individual test case results for Pass, Fail and Skipped.  It would be nice if we somehow expose the option to access individual test case results in the output json (the contents of this object)

https://github.com/EnricoMi/publish-unit-test-result-action/blob/master/python/publish/publisher.py#L296

I'm happy to submit a PR that does something like add a cleaned up version of that object to the output json if an optional parameter is supplied to the action, for instance, if it makes sense.

### PR DESCRIPTION
**Pull request by cpdeethree**
Adds the ability to optionally write out all of the individual unit test results to json file.

Intended to resolve #365

### CODE REVIEW

**Comment by cpdeethree** [comment]
Needs updates for tests, working on those currently.

**Comment by cpdeethree** [comment]
@EnricoMi let me know what you think about this approach!

File: python/test/test_publisher.py [hunk 1]
```diff
@@ -1680,16 +1746,112 @@ def test_publish_json(self):
...
+
+                    gh, gha, req, repo, commit = self.create_mocks(digest=self.base_digest, check_names=[settings.check_name])
+                    publisher = Publisher(settings, gh, gha)                    
+
                     publisher.publish_json(self.publish_data)
                     gha.error.assert_not_called()
 
                     # assert the file
                     with open(filepath, encoding='utf-8') as r:
+                        self.maxDiff = None
```
Comment (EnricoMi) [inline python/test/test_publisher.py:1854]: not needed once the tests pass, we can add this when we need to debug a test (I hope this is true for the whole project)

**Review by EnricoMi — CHANGES_REQUESTED** [review]
Very good PR, just minor changes required.

File: python/test/test_publisher.py (hunk 1 above)
Comment (cpdeethree) [reply python/test/test_publisher.py:1854]: It's not, but I can remove the one other place I saw it at as part of this PR as well

File: python/test/test_publisher.py [hunk 2]
```diff
@@ -1496,7 +1498,71 @@ def test_publish_check_with_multiple_annotation_pages(self):
             title=f'Error processing result file',
             raw_details='file'
         )],
-        check_url='http://check-run.url'
+        check_url='http://check-run.url',
+        cases=None
+    )
+
+    publish_data_with_cases = PublishData(
```
[the remark this answers is not in the document]
Comment (cpdeethree) [reply python/test/test_publisher.py]: Done!

File: python/test/test_action_script.py [hunk 3]
```diff
@@ -822,7 +824,6 @@ def test_parse_files(self):
             self.assertTrue(any([call.args[0].startswith('reading TRX files [') for call in l.debug.call_args_list]))
 
         self.assertEqual([], gha.method_calls)
-
```
[the remark this answers is not in the document]
Comment (cpdeethree) [reply python/test/test_action_script.py]: Added back in!

File: python/test/test_publisher.py [hunk 4]
```diff
@@ -1680,16 +1746,112 @@ def test_publish_json(self):
             with self.subTest(json_thousands_separator=separator):
                 with tempfile.TemporaryDirectory() as path:
                     filepath = os.path.join(path, 'file.json')
-                    settings = self.create_settings(json_file=filepath, json_thousands_separator=separator)
+                    settings = self.create_settings(json_file=filepath, json_thousands_separator=separator, json_test_case_results=False)
```
[the remark this answers is not in the document]
Comment (cpdeethree) [reply python/test/test_publisher.py]: Fixed!

File: python/test/test_publisher.py [hunk 5]
```diff
@@ -1680,16 +1746,112 @@ def test_publish_json(self):
...
+                                             "tests_skip": {"number": 6, "delta": -6}, "tests_fail": {"number": 7, "delta": -7},
+                                             "tests_error": {"number": 8, "delta": -8}, "runs": {"number": 9, "delta": -9},
+                                             "runs_succ": {"number": 10, "delta": -10},
+                                             "runs_skip": {"number": 11, "delta": -11},
+                                             "runs_fail": {"number": 12, "delta": -12},
+                                             "runs_error": {"number": 1345, "delta": -1345}, "commit": "commit",
+                                             "reference_type": "type", "reference_commit": "ref"},
+                        "annotations": 1,
+                        "check_url": "http://check-run.url",
+                        "cases": {"-keys": {
```
[the remark this answers is not in the document]
Comment (cpdeethree) [reply python/test/test_publisher.py]: Updated reduce_dict function to strip this field and not send it to the json output. Also updated the test accordingly.

**Comment by cpdeethree** [comment]
@EnricoMi I think this is ready for your eyes again. Thanks for the quick and thorough feedback!

**Comment by EnricoMi** [comment]
I have added this to the CI/CD, so this feature is used there as well: 5d01af9

Please pull from your remote git.

**Comment by cpdeethree** [comment]
> I am going to rework `class UnitTestCaseResults` today, which is difficult to serialize atm.

I suspected the failure was related to this. Thanks! Looking forward to what you come up with!

**Comment by EnricoMi** [comment]
After reworking the `UnitTestCaseResults` type, I could reproduce the error that we saw in the CI/CD, which wasn't produced by tests before. I have also beautified the structure of cases because the tuple-keys were not very meaningful and could not be JSON-serialized as they were.

I have also added an example to the README.md.

**Comment by EnricoMi** [comment]
Tests are failing due to PyPi outage: https://status.python.org/incidents/77qzc7qsb5wb

**Review by EnricoMi — APPROVED** [review]
LGTM!

**Comment by cpdeethree** [comment]
> After reworking the `UnitTestCaseResults` type, I could reproduce the error that we saw in the CI/CD, which wasn't produced by tests before. I have also beautified the structure of cases because the tuple-keys were not very meaningful and could not be JSON-serialized as they were.
> 
> I have also added an example to the README.md.

Nice work!

**Comment by cpdeethree** [comment]
@EnricoMi pypi said that they resolved the problem if you want to try retriggering the builds (I can't do it since it's my first pr here)

**Comment by EnricoMi** [comment]
Does this work for you?
https://github.com/EnricoMi/publish-unit-test-result-action/actions/runs/3313904648/jobs/5472751179#step:6:2047

**Comment by cpdeethree** [comment]
> Does this work for you? https://github.com/EnricoMi/publish-unit-test-result-action/actions/runs/3313904648/jobs/5472751179#step:6:2047

It does, that is precisely what I need. Thanks for helping quickly get this in!

**Comment by EnricoMi** [comment]
Thanks for the contribution!

**Comment by EnricoMi** [comment]
This has been released with v2.2.0, enjoy!

**Pull request merged**